core/src/sumeragi/main_loop.rs

```diff
@@ -340,18 +314,33 @@ fn commit_block(sumeragi: &Sumeragi, state: &mut State, block: impl Into<Version
     );
 
     update_topology(state, sumeragi, &committed_block);
-
     sumeragi.kura.store_block(committed_block);
-
-    cache_transaction(state, sumeragi);
+    cache_transactions(state, sumeragi);
 }
 
-fn replace_top_block(
-    sumeragi: &Sumeragi,
-    state: &mut State,
-    block: impl Into<VersionedCommittedBlock>,
-) {
-    let committed_block = block.into();
+fn replace_top_block(sumeragi: &Sumeragi, state: &mut State, block: ValidBlock) {
+    let role = state.current_topology.role(&sumeragi.peer_id);
+    let addr = &sumeragi.peer_id.address;
+
+    let header = block.header();
+    let block_hash = block.hash();
+
+    warn!(
+        %addr, %role, %block_hash,
+        peer_latest_block_hash=?state.latest_block_hash,
+        peer_latest_block_view_change_index=?state.latest_block_view_change_index,
+        consensus_latest_block_view_change_index=%header.view_change_index,
+        "Soft fork occurred: peer in inconsistent state. Rolling back and replacing top block."
+    );
+
+    // TODO: It should not be ok to commit without proxy tail signature here?
+    let committed_block = match block.commit_without_proxy_tail_signature(&state.current_topology) {
+        Ok(block) => block,
+        Err((_, err)) => {
+            error!(?err, %block_hash, "Failed to commit replacement block. Unable to resolve soft fork.");
+            return;
+        }
+    };
```

Comment:
I don't think this is correct. 

First of all why would we check `Leader` signature is present, but skip `ProxyTail` signature. 
Since keep track of all topology changes, i think we should just check that block has enough valid signatures.
Otherwise we can try to recreate proper topology from initial topology for this round + amount of view changes (it should be possible if randomness is not involved) and then do proper validation.

Second is that we are using **current** topology when we commit **previous** block.

Reply:
> I don't think this is correct.

I know, but the tests are failing if I change it to `.commit`, that is why I put the `TODO`. I think every block should have both `Leader` and `ProxyTail` signatures.